plugins/unreal/XJGame/Plugins/XJMusicPlugin/Source/XJMusicPlugin/Private/XjMusicInstanceSubsystem.cpp

```diff
@@ -0,0 +1,188 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "XjMusicInstanceSubsystem.h"
+#include "Kismet/GameplayStatics.h"
+#include <Settings/XJMusicDefaultSettings.h>
+#include <Sound/SoundBase.h>
+#include <Sound/SoundWave.h>
+#include <Engine/StreamableManager.h>
+#include <Engine/AssetManager.h>
+#include <Engine/ObjectLibrary.h>
+#include <Misc/FileHelper.h>
+#include <Runtime/Engine/Public/AudioDevice.h>
+
+#include "TimerManager.h"
+
+void UXjMusicInstanceSubsystem::RetrieveProjectsContent(const FString& Directory)
+{
+	WorldAudioDeviceHandle = GetWorld()->GetAudioDevice();
+
+	TArray<FString> WavFiles;
+
+	UXJMusicDefaultSettings* XjSettings = GetMutableDefault<UXJMusicDefaultSettings>();
+	if (!XjSettings)
+	{
+		return;
+	}
+
+	FString WorkPath = Directory;
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+
+	if (!PlatformFile.DirectoryExists(*WorkPath))
+	{
+		return;
+	}
+
+	TArray<FString> FoundAudioFilesPaths;
+
+	PlatformFile.FindFilesRecursively(FoundAudioFilesPaths, *WorkPath, *AudioExtension);
+
+	for (const FString& Path : FoundAudioFilesPaths)
+	{
+		FString Name = FPaths::GetCleanFilename(Path);
+		AudioPathsByNameLookup.Add(Name, Path);
+	}
+}
+
+void UXjMusicInstanceSubsystem::PlayAudioByName(const FString& Name, const float OverrideStartTime)
+{
+	if (SoundsMap.Contains(Name))
+	{
+		return;
+	}
+
+	USoundWave* SoundWave = GetSoundWaveByName(Name);
+	if (!SoundWave)
+	{
+		return;
+	}
+
+	if (FAudioDeviceHandle Device = GetWorld()->GetAudioDevice())
+	{
+		TSharedPtr<FActiveSound> ActiveSound = MakeShared<FActiveSound>();
+		ActiveSound->SetSound(SoundWave);
+		ActiveSound->SetWorld(GetWorld());
+
+		ActiveSound->SetVolume(1.0f);
+
+		ActiveSound->RequestedStartTime = OverrideStartTime;
+
+		ActiveSound->bIsUISound = true;
+		ActiveSound->bAllowSpatialization = false;
+
+		ActiveSound->Priority = SoundWave->GetPriority();
+		ActiveSound->SubtitlePriority = SoundWave->GetSubtitlePriority();
+
+		SoundsMap.Add(Name, ActiveSound);
+
+		Device->AddNewActiveSound(*ActiveSound.Get());
+	}
+}
+
+void UXjMusicInstanceSubsystem::StopAudioByName(const FString& Name)
+{
+	if (!SoundsMap.Contains(Name))
+	{
+		return;
+	}
+
+	FActiveSound* Sound = SoundsMap[Name].Get();
+
+	if (!Sound)
+	{
+		return;
+	}
+
+	FAudioThread::RunCommandOnAudioThread([this, Sound, Name]()
+		{
+			TArray<FActiveSound*> ActiveSounds = WorldAudioDeviceHandle->GetActiveSounds();
+
+			for (int32 SoundIndex = ActiveSounds.Num() - 1; SoundIndex >= 0; --SoundIndex)
+			{
+				FActiveSound* Inst = ActiveSounds[SoundIndex];
+				if (Inst && (Inst->GetSound()->GetName() == Sound->GetSound()->GetName()))
+				{
+					WorldAudioDeviceHandle->AddSoundToStop(Inst);
```

Comment:
The problem with this method of stopping audio is that it will just stop right now, instead of at a scheduled time. For example, imagine there is a drum loop playing called `drum_loop.wav`

1. the engine runCycle says start playing an active audio instance of `drum_loop.wav` with id 123 at chain seconds 3.0, and the current time now is 2.0 seconds. So we need to tell Unreal to play this audio starting at 3.0 seconds.
2. there is some change to the override macros and the engine decides to cut this segment short
3. the engine runCycle says that now the active audio with id 123 should get cutoff at 4.0 seconds, but the current time now is 3.0 seconds. So we need to tell Unreal to update the schedule for this audio so that it will end playback at 4.0 seconds (not now)

Reply:
This methods is only about playing audio and stopping it no scheduling. Scheduling code is located inside `PrototypeActor.cpp` where you able to find check for the EndTime to match the stop point. The reason why I doesn't implemented that here because I'm planning on converting PrototypeActor to UObject that will be spawned on scene and perform all the computations. Let me know if I misunderstand you

Reply:
I think we are talking about two different kinds of scheduling. The scheduling we need in order to get perfect sample playback timing would be internal to the audio mixer. So when we schedule an audio playback, we are telling the mixer, "start playback at X physics time, and stop playback at Y physics time" and if we did nothing more, the system would start and stop the audio on time. We should not be handling the scheduling manually.